universe/supplycommit/multi_sm_manager.go

```diff
@@ -0,0 +1,402 @@
+	if err != nil {
+		return nil, fmt.Errorf("unable to get or create state "+
+			"machine: %w", err)
+	}
+
+	return sm.RegisterStateEvents(), nil
+}
+
+// RemoveStateSub removes a state event subscriber from the state machine
+// associated with the given asset specifier. If a state machine for the asset
+// group does not exist, it will be created and started.
+func (m *MultiStateMachineManager) RemoveStateSub(assetSpec asset.Specifier,
+	sub StateSub) error {
+
+	sm, err := m.fetchStateMachine(assetSpec)
+	if err != nil {
+		return fmt.Errorf("unable to get or create state "+
+			"machine: %w", err)
+	}
+
+	sm.RemoveStateSub(sub)
+
+	return nil
+}
+
+// FetchCommitmentResp is the response type for the FetchCommitment method.
+type FetchCommitmentResp struct {
+	// SupplyTree is the supply tree for an asset. The leaves of this tree
+	// commit to the roots of the supply commit subtrees.
+	SupplyTree mssmt.Tree
+
+	// Subtrees maps a subtree type to its corresponding supply subtree.
+	Subtrees SupplyTrees
+
+	// ChainCommitment links the supply tree to its anchor transaction.
+	ChainCommitment RootCommitment
+}
+
+// FetchCommitment fetches the supply commitment for the given asset specifier.
+func (m *MultiStateMachineManager) FetchCommitment(ctx context.Context,
+	assetSpec asset.Specifier) (fn.Option[FetchCommitmentResp], error) {
+
+	var zero fn.Option[FetchCommitmentResp]
+
+	chainCommitOpt, err := m.cfg.Commitments.SupplyCommit(
+		ctx, assetSpec,
+	).Unpack()
+	if err != nil {
+		return zero, fmt.Errorf("unable to fetch supply commit: %w",
+			err)
+	}
+
+	if chainCommitOpt.IsNone() {
+		// If the chain commitment is not present, we return an empty
+		// response.
+		return zero, nil
+	}
+	chainCommit, err := chainCommitOpt.UnwrapOrErr(
+		fmt.Errorf("unable to fetch supply commit: %w", err),
+	)
+	if err != nil {
+		return zero, err
+	}
+
+	supplyTree, err := m.cfg.TreeView.FetchRootSupplyTree(
+		ctx, assetSpec,
+	).Unpack()
+	if err != nil {
+		return zero, fmt.Errorf("unable to fetch supply commit root "+
+			"supply tree: %w", err)
+	}
+
+	subtrees, err := m.cfg.TreeView.FetchSubTrees(ctx, assetSpec).Unpack()
+	if err != nil {
+		return zero, fmt.Errorf("unable to fetch supply commit sub "+
+			"trees: %w", err)
+	}
+
+	return fn.Some(FetchCommitmentResp{
+		SupplyTree:      supplyTree,
+		Subtrees:        subtrees,
+		ChainCommitment: chainCommit,
+	}), nil
+}
+
+// stateMachineCache is a thread-safe cache mapping an asset group's public key
+// to its supply commitment state machine.
+type stateMachineCache struct {
+	// mu is a mutex that is used to synchronize access to the cache.
+	mu sync.RWMutex
+
+	// cache is a map of serialized asset group public keys to their
+	// supply commitment state machines.
+	cache map[asset.SerializedKey]*StateMachine
+}
+
+// newStateMachineCache creates a new supply commit state machine cache.
+func newStateMachineCache() stateMachineCache {
+	return stateMachineCache{
```

Comment:
I'm surprised the linter didn't complain about this. But I think anything that contains a `sync.RWMutex` (or the plain one) should not be copied for safety. So this should return a pointer and it should be kept as a pointer in the parent struct.